### PR TITLE
jefftmarks/APPEALS-25002

### PR DESCRIPTION
**Pull request by jefftmarks**
Resolves [APPEALS-25002: Account for Other Means of Postponing Hearings](https://jira.devops.va.gov/browse/APPEALS-25002)

# Description
As a Caseflow developer, I need to ensure HPR tasks are cancelled whenever hearings are postponed by other means within Caseflow.

## Acceptance Criteria
- [ ] Code compiles correctly
- [ ] WHEN there is one or more open Hearing postponement request tasks for an appeal
AND one of the following occurs:
• A user submits a "Mark as complete: Postpone" action on a "No show hearing" task [Link to code](https://github.com/department-of-veterans-affairs/caseflow/blob/ba5468ec5403160d40478b1b825dad0c4d8d313f/app/models/tasks/no_show_hearing_task.rb#L54-L65)
• A user changes a hearing's disposition to "postponed" and the HearingDispositionChangeJob runs, calling AssignHearingDisposition#postpone! [Link](https://github.com/department-of-veterans-affairs/caseflow/blob/ba5468ec5403160d40478b1b825dad0c4d8d313f/app/models/tasks/assign_hearing_disposition_task.rb#L105-L111)
• A user submits a "Postpone hearing" action on a "Select hearing disposition" task from the case details page (AssignHearingDispositionTask#update_from_params) [Link](https://github.com/department-of-veterans-affairs/caseflow/blob/293dc448f051909dd5393b7398b0d8083c539a73/app/models/tasks/assign_hearing_disposition_task.rb#L81-L91)
• A user submits a "Postponed" action on a "Change hearing disposition" task (ChangeHearingDispositionTask#update_from_params) [Link](https://github.com/department-of-veterans-affairs/caseflow/blob/293dc448f051909dd5393b7398b0d8083c539a73/app/models/tasks/assign_hearing_disposition_task.rb#L81-L91)
THEN
  - [ ] Each open Hearing postponement request task for the appeal is marked cancelled
    - [ ] “Currently active tasks” is updated to remove the cancelled task
    - [ ] The Case Timeline is updated with an entry containing the following information:
      - [ ] Cancelled on: < date action submitted >
      - [ ] Event: “HearingPostponementRequestTask completed”
      - [ ] Assigned to: < Caseflow Organization or user assigned to the Hearing postponement task at the time action was submitted >
      - [ ] Assigned by: < user who assigned the Hearing postponement request task to the above Caseflow Organization or user >
      - [ ] Cancelled by: < user who submitted the action >
      - [ ] Task instructions: as specified in the attached wireframes
  - [ ] All other open hearing tasks for the appeal are cancelled
  - [ ] A new HearingTask for the appeal is created
    - [ ] The new task is assigned to Board of Veterans Appeals

## Testing Plan
1. Go to [APPEALS-29385](https://jira.devops.va.gov/browse/APPEALS-29385)
### Test Coverage
Did you include any test coverage for your code? Check below:
- [x] RSpec
- [ ] Jest
- [ ] Other

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [ ] No new code climate issues added

### CODE REVIEW

File: app/models/tasks/hearing_mail_tasks/hearing_postponement_request_mail_task.rb [hunk 1]
```diff
@@ -40,6 +40,24 @@ def available_actions(user)
...
+  # Purpose: When a hearing is postponed through the completion of a NoShowHearingTask, AssignHearingDispositionTask,
+  #          or ChangeHearingDispositionTask, cancel any open HearingPostponementRequestMailTasks in that appeal's
+  #          task tree, as the HPR mail tasks have become redundant.
+  #
+  # Params: completed_task - task object of the completed task through which the hearing was postponed
+  #         updated_at - datetime when the task was completed
+  #
+  # Return: The cancelled HPR mail tasks
+  def cancel_when_redundant(completed_task, updated_at)
+    # TO-DO when changes merged you don't need to pass in completed task, can call #open_hearing
```
[the remark this answers is not in the document]
Comment (jefftmarks) [reply app/models/tasks/hearing_mail_tasks/hearing_postponement_request_mail_task.rb:52]: Right left that in intentionally. This [PR](https://github.com/department-of-veterans-affairs/caseflow/pull/19199) is still open, but when changes are merged in I would call [this method](https://github.com/department-of-veterans-affairs/caseflow/blob/6c04997e514332ec0f0f4034181e984851ecae10/app/models/tasks/hearing_mail_tasks/hearing_postponement_request_mail_task.rb#L79-L81) instead, which would make it unnecessary to pass completed task in to grab its hearing.

File: app/controllers/appeals_controller.rb [hunk 2]
```diff
@@ -407,4 +407,3 @@ def get_appeal_object(appeals_id)
     end
   end
 end
-
```
Comment (msteele96) [inline app/controllers/appeals_controller.rb:410]: Was this due to a linting error? Otherwise, undo this change so that there is no change to this file.

File: app/models/concerns/task_extension_for_hearings.rb [hunk 3]
```diff
@@ -110,4 +110,19 @@ def withdraw_hearing(parent)
...
+  def cancel_redundant_hearing_postponement_req_tasks
+    open_hearing_postponement_requests.each { |task| task.cancel_when_redundant(self, updated_at) }
+  end
+
+  # Purpose: Finds open HearingPostponementRequestMailTasks (assigned to HearingAdmin and not MailTeam) in task tree
+  def open_hearing_postponement_requests
+    appeal.tasks.where(
+      type: HearingPostponementRequestMailTask.name,
+      assigned_to: HearingAdmin.singleton
+    ).open
```
Comment (jimfoley25) [inline app/models/concerns/task_extension_for_hearings.rb:126]: Suggest protecting against NoMethodError by changing to   )&.open

**Review by ThorntonMatthew — APPROVED** [review]
Awesome work!
